File: 342-countOfSubListInRange.py
"""_summary_
   You are given a list of integers nums and integers lo and hi. Return the number of sublists A such that lo ≤ max(A) ≤ hi.

Constraints
0 ≤ n ≤ 100,000 where n is the length of nums

Example 1
Input
nums = [1, 5, 3, 2]
lo = 1
hi = 4
Output
4

Explanation
We have the following sublists where 1 ≤ max(A) ≤ 4
[1]
[3]
[3, 2]
[2]
We can count the number of sublists that has maximum less than a upperbound R, e.g. f(R), 
and we also count the number of sublists with max less or equal than L-1 i.e. f(L-1) and the answer would be f(R)-f(L-1)
"""


def foo(nums,L,R):
    def f(upperBound):
        cur=ans=0
        for i in nums:
            if i<=upperBound:
                cur+=1
            else:
                cur=0
            ans+=cur
        return ans
    h=f(R)
    l=f(L-1)
    return f(R)-f(L-1)


def f(nums,L,R):
    ans=0
    n=len(nums)
    for i in range(n):
        for j in range(i,n):
            s=nums[i:j+1]
            if L<=max(s) and R>=max(s):
                print(s)
                ans+=1
    return ans


nums=[0,1,5,3,4]
print(f(nums,3,4))
print("----------")
print(foo(nums,3,4))

File: 342-countOfSubListInRange.cpp
#include<iostream>
#include<vector>
#include<algorithm>

using DT=std::vector<int>;

int foo(const DT& nums, int upperValue)
{
    int ans=0,cur=0;
    for(auto &v:nums)
    {
        if(v<=upperValue)
            cur+=1;
        else
            cur=0;
        ans+=cur;
    }
    return ans;
}

int countOfSubLists(const DT& nums, int L,int H)
{
    return foo(nums,H)-foo(nums,L-1);
}


int countOfSubLists_BruceForce(const DT& nums,int L,int H)
{
    int ans=0;
    if(nums.empty())
        return 0;
    
    for(int i=0;i<nums.size();++i)
    {
        DT temp;
        temp.push_back(nums[i]);
        for(int j=i;j<nums.size();++j)
        {
            temp.push_back(nums[j]);
            int max=*std::max_element(temp.begin(),temp.end());
            // int max=std::max(temp);
            if(max>=L && max<=H)
                ans+=1;
        }
    }
    return ans;
}


int main()
{
    DT nums{0,1,5,3,4};
    // int res=countOfSubLists(nums,3,4);
    int res=countOfSubLists_BruceForce(nums,3,4);
    std::cout<<res<<'\n';
    return 0;
}